ouroboros-consensus/docs/AddingAnEra.md

```diff
@@ -0,0 +1,126 @@
+`Ouroboros.Consensus.Cardano.Block`. The compiler should point the rest of the
+way. It's mostly a matter of following the existing patterns by imitating the
+previous Shelley-based case. Be sure to run both the Shelley
+(`ouroboros-consensus-shelley-test`) and the Cardano
+(`ouroboros-consensus-cardano-test`) tests.
+
+For exhaustiveness, we give an overview of the changes needed. The new era we'll
+be adding is the Alonzo era, which comes after the Mary era.
+
+## Preparation
+
+* Locate the new tag in the ledger, e.g., `AlonzoEra`. The ledger should have an
+  instance of the `ShelleyBasedEra` class (the class defined in
+  `Shelley.Spec.Ledger.API`, do not confuse it with the class with the same name
+  in consensus) for this era. This class should provide all the instances
+  consensus integration will rely on.
+
+* Note that both the affected consensus libraries and test suites are parametric
+  in the Shelley-based era or in the general era. The test suite relies on the
+  necessary `Arbitrary`, `Show`, `Eq`, ... instances to be available for all of
+  the Cardano eras. This means that a new era can only be added to the Cardano
+  eras when the ledger provides all the necessary instances for the new era,
+  *including the instances for the test suite* (e.g., the serialisation
+  generators). Note that because the test suite is parametric/generic in the
+  eras, it is not easy to temporarily omit a single era from the test suite
+  until the required instances have been added. This means that we can't extend
+  the library with the new era without extending the test suite too. In the
+  past, this has been the main blocker: the library was ready, but the required
+  instances in the ledger were missing for our test suite(s).
+
+* Add the required dependencies on the ledger package (and its test package) to
+  the `cabal.project` file. You will have to add it to some other cabal files
+  too, let the compiler tell you which ones.
+
+## `ouroboros-consensus-shelley`
+
+* Define `StandardAlonzo` in `Ouroboros.Consensus.Shelley.Eras` and add any
+  missing instances to that module, update the export list appropriately.
+
+* In `Ouroboros.Consensus.Shelley.Node`, define `ProtocolParamsAlonzo` just like
+  the existing ones. In case the era in question needs extra parameters (e.g., a
+  new genesis config?), they can be added here.
+
+## `ouroboros-consensus-shelley-test`
+
+* In `Test.Consensus.Shelley.Examples`, define an `examplesAlonzo` similar to
+  how `examplesMary` is defined. If necessary, i.e., when new type families have
+  been created, the `examples` function might have to take more arguments. This
+  is where the golden test examples are defined, but only the Shelley ones are
+  tested as part of this test suite. The others are only tested as part of the
+  `ouroboros-consensus-cardano-test` test suite.
+
+## `ouroboros-consensus-cardano`
+
+* In `Ouroboros.Consensus.Cardano.Block`, include `AlonzoEra` in `CardanoEras`
```

Comment:
Tangent: what do you think about `type CardanoEras c = ByronBlock ': ShelleyBasedEras c` or similar? Did you already weigh the pros and cons?

Reply:
I hadn't thought about that, good idea! I don't see any real cons (an extra indirection :roll_eyes:). I can push a commit.

Reply:
Ah, I just realised why we didn't do that. Compare for example Allegra: in `CardanoEras`  it's `ShelleyBlock (AllegraEra c)`, in `ShelleyBasedEras` it's `AllegraEra c`. This means we'd need a type-level `map`, e.g.
```haskell
type CardanoEras c = ByronBlock ': Map ShelleyBlock (ShelleyBasedEras c)
```
For simplicity (there are already too many people that don't understand type families, even though here it would just be an almost literate promotion from the value level to the type level), we didn't bother.

For reference:
```haskell
type CardanoEras c = ByronBlock ': Map ShelleyBlock (ShelleyBasedEras c)

type ShelleyBasedEras c = '[ShelleyEra c, AllegraEra c, MaryEra c]

type family Map f xs where
  Map _ '[]       = '[]
  Map f (x ': xs) = f x ': Map f xs
```

Reply:
FWIW, the core concepts in `generics-sop` is parameterized over functors precisely to avoid working with type-level maps, as in practice that can get quite painful.

Reply:
Yep, I missed that. I agree, with your decision. Maybe in another 5 years `Map` will be a non-issue...